File: interview/intern_coding/papertests/hwy20240417/B.java
package interview.papertests.hwy20240417;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

public class B {
    static int ans = 0;
    static int m;
    static Map<String, List<String>> children = new HashMap<>();
    static Map<String, Integer> stat1 = new HashMap<>();
    static Map<String, Integer> stat2 = new HashMap<>();
    static Map<String, Boolean> cloud = new HashMap<>();

    public static void main(String[] args) {
        String input = "40 12\n" +
                "a * 0 2\n" +
                "a * 1 2\n" +
                "b a 0 3\n" +
                "b a 1 5\n" +
                "c a 1 3\n" +
                "d a 0 1\n" +
                "d a 1 3\n" +
                "e b 0 2\n" +
                "f * 0 8\n" +
                "f * 1 10\n" +
                "g f 1 2\n" +
                "h * 0 4";

        run(System.in, System.out, input);
    }

    public static void run(InputStream inputStream, OutputStream outputStream, String input) {
        Scanner scanner = new Scanner(input);
        PrintWriter out = new PrintWriter(outputStream);
        m = scanner.nextInt();
        int n = scanner.nextInt();
        scanner.nextLine(); // Consume newline


        for (int i = 0; i < n; i++) {
            String line = scanner.nextLine();
            String[] parts = line.split("\\s+");
            String a = parts[0];
            String b = parts[1];
            int c = Integer.parseInt(parts[2]);
            int d = Integer.parseInt(parts[3]);

            if (!b.equals("*")) {
                if (!children.containsKey(b)) {
                    children.put(b, new ArrayList<>());
                }
                children.get(b).add(a);
            } else {
                cloud.put(a, true);
            }

            if (c == 0) {
                stat1.put(a, d);
            } else {
                stat2.put(a, d);
            }
        }

        for (String k : cloud.keySet()) {
            if (cloud.get(k)) {
                dfs(k);
            }
        }

        out.println(ans);
        out.flush();
    }

    private static int[] dfs(String i) {
        int[] res = new int[2];

        for (String j : children.getOrDefault(i, new ArrayList<>())) {
            int[] sub = dfs(j);
            res[0] += sub[0];
            res[1] += sub[1];
        }

        res[0] += stat1.getOrDefault(i, 0);
        res[1] += stat2.getOrDefault(i, 0);
        if (5*res[0]+2*res[1] > m && cloud.getOrDefault(i, false)) {
            ans ++;
        }
        return res;
    }
}

File: interview/intern_coding/papertests/hwy20240417/b.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
	// "strings"
)

// type info struct {n1, n2 int}


func main() {
	
	run(os.Stdin, os.Stdout)
}
/*
40 12
a * 0 2
a * 1 2
b a 0 3
b a 1 5
c a 1 3
d a 0 1
d a 1 3
e b 0 2
f * 0 8
f * 1 10
g f 1 2
h * 0 4
*/ 
func run(rd io.Reader, wt io.Writer) {
	in := bufio.NewReader(rd)
	out := bufio.NewWriter(wt)
	defer out.Flush()
	
	var m, n int
	Fscan(in, &m, &n)
	var t string
	Fscan(in, &t)
	children := map[string][]string{}
	stat1 := map[string]int{}
	stat2 := map[string]int{}
	// 无父节点为云服务
	cloud := map[string]bool{}
	for i:=0; i < n; i ++ {
		var a, b string
		var c, d int
		Fscanf(in, "%s %s %d %d\n", &a, &b, &c, &d)
		Fprintln(out, a, b, c, d)
		if b != "*" {
			children[b] = append(children[b], a)
		} else {
			cloud[a] = true
		}
		if c == 0 {
			stat1[a] = d
		} else {
			stat2[a] = d
		}
	}
	ans := 0
	var dfs func(string) [2]int
	dfs = func(i string) (res [2]int) {
		
		for _, j := range children[i] {
			sub := dfs(j)
			res[0] += sub[0]
			res[1] += sub[1]
		}
		res[0] += stat1[i]
		res[1] += stat2[i]
		if 5*res[0]+2*res[1] > m {
			ans ++
		}
		return res

	}
	for k, v := range cloud {
		if v {
			dfs(k)
		}
	}

	Fprintln(out, ans)


}